xi_test: LSB test framework

sruon opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

Courtesy of Sora

> <img width="256" height="384" alt="jgvtqfessrc2t9c70zj8pt" src="https://github.com/user-attachments/assets/ad0714bd-7e01-4feb-bd99-b9a51264b860" />

## Forewords
I'll acknowledge upfront that the bulk is strongly inspired, if not copy-pasted from Eden implementation and I cannot thank you enough for it.

I suggest reviewing the PR in this order:
- Pull the PR, build and run xi_test
  - Don't need any specific argument, but `xi_test --keep-going --output test.ctrf.json` is what I iterate with
- Open one test and go over the syntax
  - `scripts/tests/missions/cop.lua` is a good representation of mission tests
  - `scripts/tests/jobs/war/abilities/berserk.lua` is a simple example of how to test job abilities
  - `scripts/tests/systems/spells/weather.lua` is a good representation on how to test complex systems with spies
  - `scripts/tests/framework` contain an (almost) full set of tests representing framework capabilities
- If you don't like the syntax, complexity and/or it didn't make you want to write tests, then **stop here and provide that feedback.**
- If you make it there, read the code and (possibly) scream

## Quick view
New process: **xi_test**, follows this flow:
- Collects all lua scripts in `scripts/tests` and read them to load all test suites
- Filter suites and cases according to parameters passed to `xi_test` (supports positive/negative matching and regex on both files and test cases, super flexible!)
- Create a new `MapEngine` and `WorldEngine` with special parameters to only initialize what matters
- Executes **synchronously** each test case in dedicated lua environment
  - The environment contains a set of global functions and classes **only available** to the test environment
- Push results to various reporters (console + CTRF supported atm)
- Exits 0/1 depending on test failures

The framework is heavily inspired by:
- [busted](https://lunarmodules.github.io/busted/) for test definitions
- [luassert](https://github.com/lunarmodules/luassert) for spies/stubs and general assertions
- [pytest](https://docs.pytest.org/en/stable/) in general
- [gtest](https://github.com/google/googletest) for the console reporter

## Features
- Test lifecycle functions: `setup/teardown/before_each/after_each`
- Flexible filename and test cases matching
- In-memory logger
- Souped-up `CLuaBaseEntity` with common assertions wired in and helpers to emit packets/complete actions
  - Common pre-canned packets exposed as helper
  - Can drop down to FFI to emit unimplemented packets and/or that require deeper customization without writing an extra binding
- Mobs are also assertable (to a lesser degree)
- Control over the PRNG seed
- Console and CTRF JSON reporter
- (Some) Sandboxing in between suites
- Spies and Stubs as test doubles to observe and/or replace global functions!
  - Assertable and automatically reverted on case ends
- Decent set of tests (most of it courtesy of Eden)
  - Nearly all of CoP/RoZ missions
  - Some of Bastok/Sandy missions
  - Framework tests
  - Couple tests to showcase usage

## Differences with #7670 
### No extra dependencies
Did the one thing I didn't initially want to do and wrote a whole front-end. 

The only extra C++ dependency is `termcolor`, a header only console coloring library because I really don't want to handle ANSI escape codes by hand

No more luarocks, no more busted, no more luassert. It's all natural, grass-fed code in size I hope will prove maintainable.

### Everything is implemented C++ side
Got rid of all the various lua files that made up the original framework. 

While there was nothing wrong with the approach, I quickly encountered patterns where I was crossing in between C++ and Lua _too many times_ depending on what I needed to do. 

I felt standardizing on one side was the better long-term approach.

### Merged client and test entity into ~~CTestPuppet~~ CClientEntityPair!
One common pattern when writing tests used to be `client, player = xi.test.world:spawnPlayer()`, where player is a `CBaseEntity` and client was a new object destined to hold methods representing an actual human interfacing with a FFXI game client.

While the distinction is important to keep in mind, I felt it didn't add anything of value to the tests and made them less readable. 

Thus comes  `CClientEntityPair`, an assertable `CBaseEntity` representing the current character with a bunch of helpers wired in to perform client-like interactions. Each type of helper is maintained in separate classes so as to keep the whole thing relatively easy to read.

### Improved sandboxing
Every test suite gets a nested copy-on-write environment, courtesy of sol2. The root suite is inheriting from the global lua environment, its child suites inheriting from its environment and so on.

This is not a perfect sandbox and does not guarantee a test cannot insert/remove entries from global tables but it's _something_.

Environments are shared across all test cases in a suite.

### Ticking no longer interacts with the TaskManager
I tried to be smart and just move the clock forward and _pray_ the `CTaskManager` would run the tasks I need. Unfortunately this has proven to be incredibly hard to dial correctly. 

Instead the `xi.test.world:tick()` func now takes an enum specifying which tasks you want to execute and will move the clock forward + call the tasks directly. 

Yes it's probably not correctly named if we're being pedantic but it works and you don't really need to know about the specifics unless you are testing very specific systems.

## Tangentially related
### Expanded Arguments configuration
They can now take an enum specify what _sort_ of argument they are: single, flag, multiple. argparse then registers them correctly accordingly.

### Reverted former IF changes
My previous change was based on an incorrect understanding of IF. I thought the zones and NPCs needed to be loaded before it could insert the containers and I have since come to understand it doesn't exactly matter.

- Reverts previous changes to the main entrypoint
- Adds an _alternate_ entrypoint that takes a list of zones slightly differently (map of ID/names)
- Moved IF init _out of_ `LoadZones` and into `LoadZoneList` 
  - makes no difference for actual runs of xi_map, but it does for xi_test

### New SOL_ macros
- Introduces 'SOL_USERTYPE_INHERIT' to register a new bound class that inherits from one or more classes.
  - This is used to build the relationship between CLuaBaseEntity > CLuaTestEntity > CLuaClientEntityPair
- Introduces 'SOL_READONLY' to register read-only properties (through getters only atm). 
  - This is actually something I believe we should be using more in-lieu of writing ad-hoc bindings.

## Jank
- Several places where I used either lua bindings directly from C++ or the actual underlying CCharEntity method, depending on what I needed.
  - While it stems from me just stumbling my way through, I actually believe using the lua bindings from C++ is legitimate given `xi_test` status as a test framework. 
  - I don't particularly care about how the information is retrieved and the lua bindings provide a degree of separation from the implementation that would probably prove more maintainable over the long run.
- `TestChar/CLuaClientEntityPair` design/separation does not feel fleshed out enough but that's the best I can deliver _right now_.
- Someone is probably going to scream at the sheer amount of new Lua classes but I prefer that over something like `CLuaBaseEntity` with hundreds of bindings.
- There's a little bit of code in the CTRF report type header that only exists because C++ doesn't like having a struct member named `stdout`
- There's a little bit of jank related to the fact `CCharEntity` are destroyed in between zones. 
  - To the best of my knowledge Eden framework did not have this limitation but I don't have a better idea right now.
  - The key constraint is being able to reuse the same `CBaseEntity` across multiple zones, where the underlying pointer is going to change. What I propose works but it doesn't feel _exactly right_  to me at the moment.

## Remaining
- [x] MacOS not building with some sol::nil error
- [x] Dialing in how failures are reported/presented (see https://github.com/ctrf-io/github-test-reporter?tab=readme-ov-file#report-showcase)
- [x] Windows CI job
- [x] CoP sacrarium test is flaky
- [x] Add a handful of tests that were requested in private discussions
- [x] Publish documentation on the wiki
- [x] Little bit of testing of the actual processes but I did not _change much_
- [x] Eden co-authors

## What would come in future iterations
- Code coverage
- Helpers to write standardized tests for IF/Missions
- `--watch` mode
- Possibly the ability to run several tests in parallel but that requires careful consideration
  - The new scheduler may shine here...
- Tracy support
- General optimization
- Proper mock support, with front-loaded assertions
  - I suspect the spy/stub capability is good enough but I'd love to be proven wrong
- Modules support
- Map to world communication
- Multi-map support
- Auto-generate FFI definitions

sruon committed:
xi_test: LSB test framework

Co-Authored-By: KnowOne134 <35616771+KnowOne134@users.noreply.github.com>
Co-Authored-By: InoUno <774909+InoUno@users.noreply.github.com>
Co-Authored-By: Corey <2593549+cocosolos@users.noreply.github.com>

zach2good commented on .github/workflows/build.yml:
```diff
@@ -620,6 +620,70 @@ jobs:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
+          submodules: recursive # For navmeshes
```
I think it's a good idea for us to do this here, but not in any of the other CI jobs - I've always been under the impression that this is computationally and bandwidth heavy. If we have workloads where we want to verify that incoming additions to spawn points and pathing are valid, we can now make them inside xi_test 👍

zach2good commented on .github/workflows/build.yml:
```diff
@@ -620,6 +620,70 @@ jobs:
+      - name: Copy settings
+        run: |
+          cp settings/default/* settings/
+      - name: Run xi_test
+        run: |
+          chmod +x xi_test
+          ./xi_test --keep-going --output tests.ctrf.json
+      - name: Publish Test Report
+        uses: ctrf-io/github-test-reporter@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          report-path: './tests.ctrf.json'
+          github-report: true
+          summary: true
+          pull-request: false
+          status-check: true
+          status-check-name: 'xi_test (Linux)'
+          annotate: true
+          upload-artifact: true
+          artifact-name: 'xi_test-ctrf-linux'
+        if: always()
```
👍

zach2good commented on scripts/specs/test/CTestPuppet.lua:
```diff
@@ -0,0 +1,59 @@
+---@meta
+
+-- Test puppet that combines client and entity functionality
```
I feel like `puppet` is an overloaded term in the context of this codebase. My first thought was "why do automatons need their own thing". Can we come up with something better? Even if it's verbose like `ClientEntityPair` (just an example)

zach2good commented on scripts/tests/framework/failures/test_describe_failure.lua:
```diff
@@ -0,0 +1,18 @@
+-- This file is used to test framework error handling behavior
```
Is there a scenario where we want to conditionally run these before the test suite starts to verify the test infrastructure actually still works?

sruon replied:
The logical next step would be to introduce some sort of C++ level testing and test the framework with it. 

Right now the framework is not geared to _expect_ failures.

zach2good commented on scripts/tests/framework/failures/test_execute_failure.lua:
```diff
@@ -0,0 +1,10 @@
+-- Uncomment to verify that the test framework properly handles describe block failures
+
+describe('Bad test', function()
+    it('should fail', function()
```
👍

zach2good commented on scripts/tests/framework/lifecycle.lua:
```diff
@@ -0,0 +1,130 @@
+describe('Test lifecycle', function()
+    describe('setup()', function()
```
These are good examples 👍

zach2good commented on scripts/tests/framework/prng.lua:
```diff
@@ -0,0 +1,36 @@
+        local expected = { 134, 137, 452, 22, 351, 912, 471, 75, 570, 636 }
+        local actual   = {}
+
+        xi.test.world:setSeed(1)
```
cool!

zach2good commented on scripts/tests/framework/test_double.lua:
```diff
@@ -0,0 +1,77 @@
+        it('can replace global functions with a new implementation', function()
+            player:setSkillLevel(xi.skill.BONECRAFT, 300)
+
+            local s = stub('xi.crafting.getRealSkill', function()
```
Super useful!

zach2good commented on scripts/tests/jobs/smn/perpetuation.lua:
```diff
@@ -0,0 +1,229 @@
+        player:despawnPet()
+        player:spawnPet(xi.petId.CARBUNCLE)
+        verifyPerpetuationCost(player, 9)
+    end)
```
and then presumably the whole world comes down and all the entities and clients are cleaned up?

sruon replied:
Not sure if you were specifically referencing this test or in general. 

The clients are scoped to a single file and cleaned up after each top level suite. The 'world' itself is only initialized once and then reused across files. 

I have not yet seen issues with reusing the world but it's on my mind as something to look out for as we should probably roll it back to some sort of initial checkpoint _somehow_. It would be simple to recreate it on each test but it is computationally expensive to do so.

I'll probably think about it some more when I add support for multiple MapEngine.

zach2good commented on scripts/tests/jobs/smn/perpetuation.lua:
```diff
@@ -0,0 +1,229 @@
+        verifyPerpetuationCost(player, 9)
+    end)
+
+    it('is reduced by elemental staves', function()
```
Yeah, this syntax is nice 👍

zach2good commented on scripts/tests/missions/bastok.lua:
```diff
@@ -0,0 +1,171 @@
+
+            -- Pick up mission from gate guard.
+            puppet:gotoZone(xi.zone.BASTOK_MARKETS)
+            puppet.entities:gotoAndTrigger('Cleades', { eventId = 1000, finishOption = 0 })
```
I would expect this to be: `puppet:gotoAndTrigger(...`, with a more obvious verify/expects payload. I haven't looked at how this is laid out/bound yet though

zach2good replied:
Or even:

puppet:gotoAndTrigger('Name') -> returns the table

```lua
assertEQ(puppet:gotoAndTrigger('Name'), { eventId = 120 })
```

sruon replied:
I don't have a better design to offer right this minute but one thing I'm gearing towards is emulating the way we do quests/missions, where we simply define a table of handlers mapped to different sections and somehow the framework wires it all up.

I believe we could offer a similar model where each quest/mission/NPC-interaction-driven test is just a handful of very simple sections in a table, then you can drop down to specialized APIs if you really need to do fancy stuff.

zach2good commented on src/map/lua/luautils.cpp:
```diff
@@ -1227,16 +1228,23 @@ namespace luautils
         if (!result.valid())
         {
-            sol::error err = result;
+            const sol::error err = result;
```
👍

zach2good commented on src/map/map_networking.cpp:
```diff
@@ -71,6 +71,12 @@ MapNetworking::MapNetworking(MapStatistics& mapStatistics, const MapConfig& mapC
 {
     TracyZoneScoped;
 
+    // Embedded map server for testing does not actually need to open a socket
```
I haven't looked at the guts for this, but presumably we still need this but it's just inactive? Otherwise I think it would make sense to have an interface for MapNetworking which this fulfils, and then we have a TestMapNetworking which we can poke and prod during testing - which doesn't actually do packet things, etc.

sruon replied:
As of right now I'm cheating and injecting packets straight into the packet handlers. 

I'd like to eventually test the network layer but I'm gonna need a little time to think about it, possibly when I add proper Map<>World / MultiMap support.

zach2good commented on src/test/enums/test_status.h:
```diff
@@ -0,0 +1,29 @@
+{
+    Passed,
+    Failed,
+    Skipped
```
Trailing comma please

zach2good commented on src/test/in_memory_sink.h:
```diff
@@ -0,0 +1,56 @@
+#include <string>
+#include <vector>
+
+class InMemorySink final : public spdlog::sinks::base_sink<std::mutex>
```
👍

WinterSolstice8 commented on scripts/tests/jobs/thf/traits/treasure_hunter.lua:
```diff
@@ -0,0 +1,91 @@
+            {
+                -- item, expected drop rate, tolerance
+                -- TODO: Tolerance needs to be tighter
+                [17489925] =
```
hardcoded ID here, not ideal but we can fix it later

WinterSolstice8 commented on scripts/tests/missions/cop.lua:
```diff
@@ -0,0 +1,1127 @@
+            player:gotoZone(xi.zone.ALTAIEU)
+            player.entities:gotoAndTrigger('_0x0', { eventId = 164 })
+
+            local southMob1 = player.entities:get(16912829)
```
More hardcodes but as mentioned before can fix them later

WinterSolstice8 commented on settings/default/test.lua:
```diff
@@ -11,4 +11,5 @@ xi.settings = xi.settings or {}
 
 xi.settings.test =
 {
+    PATTERN = "[%T][%^%-4!l%$][%*] %v",
```
Comment for this? quite cryptic

WinterSolstice8 commented on src/map/utils/zoneutils.cpp:
```diff
@@ -109,7 +109,6 @@ namespace zoneutils
             return g_PZoneList.at(zoneId);
         }
 
-        ShowWarning(fmt::format("Invalid zone requested: {}", zoneId));
```
This thing is super noisy and is meaningless in a multi-process setup so good riddance

All it means is the zoneId isn't loaded on the current process

WinterSolstice8 reviewed (approved):
Don't know what to say other than exceptional work

this will no doubt be a PR to be remembered for the ages

sruon commented:
- Made assertions chainable

```lua
player.assert
    :hasEffect(xi.effect.BERSERK)
    :hasModifier(xi.mod.ATTP, 25)

player.assert
    :hasCompletedQuest(xi.questLog.BASTOK, xi.quest.id.bastok.SHADY_BUSINESS)
    :hasGil(350) -- Quest reward is 350 gil
```

- Added a comment describing the logging pattern
- Added xi_test Linux executable to gitignore
- Added a very light Gilfinder test
- Updated wishlist in #7986 
- Opened https://github.com/LandSandBoat/lsb-wiki/pull/109 for wiki doc
  - I'll probably iterate once more on it
